tensorflow_lite_support/ios/task/processor/sources/TFLSegmentationResult.h

```diff
@@ -0,0 +1,97 @@
+/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================*/
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/** Holds a label associated with an RGB color, for display purposes. */
+@interface TFLColoredLabel : NSObject
+
+/** The RGB color components for the label, in the [0, 255] range. */
+@property(nonatomic, assign) UInt8 r;
+@property(nonatomic, assign) UInt8 g;
+@property(nonatomic, assign) UInt8 b;
+
+/** The class name, as provided in the label map packed in the TFLite Model
+ * Metadata.
+ */
+@property(nonatomic, copy) NSString *label;
+
+/** The display name, as provided in the label map (if available) packed in
+ * the TFLite Model Metadata. See `display_names_locale` field in
+ * ImageSegmenterOptions.
+ */
+@property(nonatomic, copy) NSString *displayName;
+
+@end
+
+/** Encapsulates a resulting segmentation mask and associated metadata. */
+@interface TFLSegmentation : NSObject
+
+/**
+ * The width of the mask. This is an intrinsic parameter of the model being
+ * used, and does not depend on the input image dimensions.
+ */
+@property(nonatomic, assign) UInt8 width;
+
+/**
+ *  The height of the mask. This is an intrinsic parameter of the model being
+ * used, and does not depend on the input image dimensions.
+ */
+@property(nonatomic, assign) UInt8 height;
+
+/**
+ *  One confidence masks of size `width` x `height` for each of the supported
+ * classes. The value of each pixel in these masks represents the confidence
+ * score for this particular class.
+ */
+@property(nonatomic, assign) float *_Nullable *_Nullable confidenceMasks;
```

Comment:
I have not implemented swift tests. But from my understanding this should translate to [[float]] in Swift. The reason I used this approach is because of the overhead of NSArrays for floats. Floats will have to be wrapped in NSNumbers. So for things like these I saw general recommendation that float arrays should be used directly. The only downside is that I had to alloc these arrays on creation and free them in dealloc in objc since ARC won't take care of it like all C stuff we are doing in ObjC. But that will be handled by our library and ObjC users won't have to worry about it.
But if the overhead won't be a problem then we can use NSArray. In any case if Swift tests don't translate the way we expected, I'll change it to an NSArray.

Reply:
Yeah it seems that if we use NSArray, we'll need to do memory copy of values in the mask, which isn't ideal ([reference](https://stackoverflow.com/questions/57228532/how-can-i-send-and-receive-extremely-large-floating-arrays-from-swift-to-objec)). Float array in ObjC does seem to be bridged to [Float] in Swift so that shouldn't be a problem too.

To make sure that we don't miss anything while reviewing this API, can you finish the ObjC implementation of this ImageSegmenter API, write a simple Swift and ObjC test for it, and commit the whole implementation to a branch in khanhlvg/tflite-support repo? That will give me a better understanding about the whole implementation to review this PR?

Reply:
I have committed m .m files [here](https://github.com/khanhlvg/tflite-support/tree/ios-image-segmentation) . Haven't updated tests. I will let you know once I commit them.

Reply:
I have updated the Swift tests. Category Mask its an `UnsafeMutablePointer<UInt8>?` and confidence masks is `UnsafeMutablePointer<UnsafeMutablePointer<Float>?>?` in swift. This is the common datatype for things like pixel data. So we should be fine. And [usage](https://github.com/khanhlvg/tflite-support/blob/ios-image-segmentation/tensorflow_lite_support/ios/test/task/vision/image_segmenter/TFLImageSegmenterTests.swift#L71) is very simple for category mask. You can check lines below it. Also we will 2 loops to scan through the confidence masks as usual.

Reply:
Thanks Prianka. I think this API is a bit difficult to understand for a typical iOS developer so I think we should create custom Objective C object type to store the masks. In particular:

```objc
@interface TFLConfidenceMask : NSObject

@property (nonatomic, assign) float* mask;
@property (nonatomic, assign) NSUInteger width;
@property (nonatomic, assign) NSUInteger height;

@end

@interface TFLCategoryMask : NSObject

@property (nonatomic, assign) NSUInteger* mask;
@property (nonatomic, assign) NSUInteger width;
@property (nonatomic, assign) NSUInteger height;

@end

@interface TFLSegmentation : NSObject

// ...

@property (nonatomic, strong) NSArray<TFLConfidenceMask *> *confidenceMask;

@property (nonatomic, strong) TFLConfidenceMask *categoryMask;

// ...
@end
```

@lu-wang-g @mbrenon WDYT about having the ObjC object structure to be slight different to the C++ proto to make it easier to understand for iOS developers? I roughly follow the structure of the Java API, but as we don't have the equivalent of TensorImage in ObjC, so I need to create a new class.

I also did a benchmark to compare the `NSArray<NSNumber *> ` performance to `float *` and found that the float array is way faster, so let's stick with it. For the record, I did two benchmarks on iPhone 14 SE.
1. Benchmark 1: 
* Copy a 2D C array (size: 1024 * 1024) to a NSArray<NSArray<NSNumber*>*>*: 134ms
* Loop through the array in Swift and add all elements: 169ms
2. Benchmark 2: 
* Copy a 2D C array (size: 1024 * 1024) to a float** array: 0.7ms
* Loop through the array in Swift and add all elements: 1.2ms

Reply:
The benchmark results are quite staggering. This option LGTM: I think we want something that's idiomatic for iOS developers while preserving performance, even if it means not sticking strictly to the C++ API.

Reply:
IMO sticking to C types are fine for segmentation masks since iOS users who deal with pixel data are frequently exposed to C pointers considering the performance gap.

Reply:
khanh's proposal looks good to me. But still wanna check, is it possible to use ODML Image library to store the mask? I didn't look deep into the code, but ideally, we should use ODML Image to store any image data. It will improve our interoperability with other frameworks.

Reply:
@lu-wang-g Unlike Java's TensorImage, the [GMLImage](https://github.com/tensorflow/tflite-support/blob/master/tensorflow_lite_support/odml/ios/image/sources/GMLImage.m) class in the ODML Image Library is just a holder for the actual image storage instance (can be either UIImage, CVImageBuffer, CVPixelBuffer), and it's the responsible of the user to convert it to the format they need (e.g. RGB float array). If we return a segmentation mask as GMLImage, we'll need to use CVPixelBuffer as the underlying storage format. The code to use the segmentation mark will require extracting the CVPixelBuffer from the GMLImage instance, and process it 
 like this MLKit sample:
https://github.com/googlesamples/mlkit/blob/4be2fe856e0b3fd95af370fe5413a33494698067/ios/quickstarts/vision/VisionExample/UIUtilities.swift#L130

However, if we return the segmentation map as a float* or NSInteger* array, user will only need to do a nested for loop to get the value (~3 lines of code). Therefore, I'd suggest we create a custom data type instead of using GMLImage, as it'll make it even more complex than without it.

@priankakariatyml I agree that iOS users can deal with float / integer pointers. However, I found returning the confidenceMasks and categoryMask directly in the TFLSegmentation object confusing because confidenceMasks is a 1D pointer array of a 2D-flatten pointer array, while categoryMask is a 2D-flatten pointer array. I want to unify the definition of masks by creating `TFLConfidenceMask` and `TFLCategoryMask ` so that the mask values always stored as 2D-flatten pointer array, and make it clear that confidenceMasks contains an NSArray of  TFLConfidenceMask. 

I did another round of benchmark to compare if adding TFLConfidenceMask as a wrapper around float** added significant overhead. It turned out that for **100 masks** of 1024 * 1024 size, using TFLConfidenceMask is even slightly faster.
```
TFLConfidenceMask: create (ms) 48.526125
TFLConfidenceMask: read (ms) 125.555834

float 1D-array: create (ms) 49.74275
float 1D-array: read (ms) 127.487
```

I think we can be confident to proceed with creating TFLCategoryMask and TFLConfidenceMask.